source/common/router/router.cc

```diff
@@ -525,61 +525,46 @@ void Filter::onResponseTimeout() {
   ENVOY_STREAM_LOG(debug, "upstream timeout", *callbacks_);
   cluster_->stats().upstream_rq_timeout_.inc();
 
-  // It's possible to timeout during a retry backoff delay when we have no upstream request. In
-  // this case we fake a reset since onUpstreamReset() doesn't care.
+  // It's possible to timeout during a retry backoff delay when we have no upstream request.
   if (upstream_request_) {
     if (upstream_request_->upstream_host_) {
       upstream_request_->upstream_host_->stats().rq_timeout_.inc();
     }
     upstream_request_->resetStream();
   }
 
-  onUpstreamReset(UpstreamResetType::GlobalTimeout, absl::optional<Http::StreamResetReason>());
+  updateOutlierDetection(timeout_response_code_);
+  const absl::string_view body =
+      timeout_response_code_ == Http::Code::GatewayTimeout ? "upstream request timeout" : "";
+  onUpstreamAbort(timeout_response_code_, StreamInfo::ResponseFlag::UpstreamRequestTimeout, body,
+                  false);
 }
 
-void Filter::onUpstreamReset(UpstreamResetType type,
-                             const absl::optional<Http::StreamResetReason> reset_reason) {
-  ASSERT(type == UpstreamResetType::GlobalTimeout || upstream_request_);
-  if (type == UpstreamResetType::Reset) {
-    ENVOY_STREAM_LOG(debug, "upstream reset: reset reason {}", *callbacks_,
-                     reset_reason ? Http::Utility::resetReasonToString(reset_reason.value()) : "");
+void Filter::onPerTryTimeout() {
+  updateOutlierDetection(timeout_response_code_);
+
+  if (maybeRetryReset(Http::StreamResetReason::LocalReset)) {
+    return;
   }
 
+  const absl::string_view body =
```

Comment:
Can you fold this code into `onUpstreamTimeoutAbort()` or similar given that it's identical with the global timeout case above and does error message building, etc.?

Reply:
that seems a bit weird to me since i'd basically conditionally ignore the `body` argument depending on HTTP code. Maybe I can move the body creation into a `bodyForTimeout` helper to dry this?

Reply:
👍